### PR TITLE
Subclass existentials - PrintAsObjC support

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/PrintAsObjC/PrintAsObjC.cpp [hunk 1]
```diff
@@ -1579,62 +1580,65 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
...
-      if (*knownKind == KnownProtocolKind::AnyObject) {
-        printNullability(optionalKind);
-        return;
+    if (layout.superclass) {
+      auto *CD = layout.superclass->getClassOrBoundGenericClass();
+      if (isMetatype) {
+        if (CD->isObjC())
+          os << "SWIFT_METATYPE(" << getNameForObjC(CD) << ")";
+        else
+          os << "Class";
```
Comment (jrose-apple) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: When can this happen?

Comment (slavapestov) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: Probably never, o was cargo culting the Metatype case where we seem to print non-@objc metatypes as Class.

Comment (jrose-apple) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: Ah, I didn't remember that was there. I suspect it's not doing anything there either at this point, but it's fine to not investigate that right now.

File: lib/PrintAsObjC/PrintAsObjC.cpp [hunk 2]
```diff
@@ -1579,62 +1580,65 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
...
+      } else {
+        os << getNameForObjC(CD);
+        if (auto *BGT = layout.superclass->getAs<BoundGenericClassType>()) {
+          os << '<';
+          print(BGT->getGenericArgs()[0], None);
```
Comment (DougGregor) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: You can use the `interleave` algorithm here...

**Review by DougGregor — APPROVED** [review]
LGTM

File: lib/PrintAsObjC/PrintAsObjC.cpp (hunk 2 above)
Comment (slavapestov) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: Ah, good idea. This was a bit of copy and paste on my part so I cleaned it up and factored it out into a new method.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Pull request merged**